di/_utils/concurrency.py

```diff
@@ -0,0 +1,120 @@
+import inspect
+from contextlib import AsyncExitStack, asynccontextmanager, contextmanager
+from functools import partial, wraps
+from typing import (
+    Any,
+    AsyncGenerator,
+    AsyncIterator,
+    Awaitable,
+    Callable,
+    ContextManager,
+    Iterator,
+    Optional,
+    ParamSpec,
+    TypeVar,
+    Union,
+    cast,
+    overload,
+)
+
+import anyio
+from anyio.streams.memory import MemoryObjectReceiveStream, MemoryObjectSendStream
+
+T = TypeVar("T")
+
+
+def cm_thead_worker(
+    cm: ContextManager[T],
+    res_stream: MemoryObjectSendStream[T],
+    err_stream: MemoryObjectReceiveStream[Optional[Exception]],
+) -> None:
+    with cm as res:
+        anyio.from_thread.run(res_stream.send, res)
+        exc = anyio.from_thread.run(err_stream.receive)
+        if exc:
+            raise exc
+
+
+@asynccontextmanager
+async def contextmanager_in_threadpool(
+    cm: ContextManager[T],
+) -> AsyncGenerator[T, None]:
+    # streams for the data
+    send_res, rcv_res = anyio.create_memory_object_stream(  # type: ignore
+        float("inf"), item_type=Any
+    )
+    # streams for exceptions
+    send_err, rcv_err = anyio.create_memory_object_stream(  # type: ignore
+        float("inf"), item_type=Optional[Exception]
+    )
+
+    async with AsyncExitStack() as stack:
+        stack.enter_context(rcv_res)
+        stack.enter_context(rcv_err)
+        stack.enter_context(send_res)
+        stack.enter_context(send_err)
+        async with anyio.create_task_group() as tg:
+            tg.start_soon(
```

Comment:
This task group runs into issues with this other TaskGroup:

https://github.com/adriangb/di/blob/76444e7e6488569346bc06368fdabfc597bc33b5/di/executors/_async.py#L60

Currently `di` runs setups using the TaskGroup linked above but teardown do not run in a TaskGroup. So really _any_ async context manager that internally has a cancel scope that straddles the `yield` will run into issues 😞

Reply:
We could just not run context managers in threads. But on the other hand being able to parse the request body while we wait for a database connection or get JWK keys to validate auth seems damn useful.

Reply:
Guess we could mark certain context managers as "task safe"? We'd make that the default. Then this context manager and others would get some sort of marker?